static-checks/ansible/allowed-modules/test.py

```diff
@@ -0,0 +1,134 @@
+
+    Use as a context manager, returning a filesystem path to the
+    modified datastream.
+    """
+    exprs = set()
+    exprs.add(re.compile(r'<.*Rule.*id="'))
+    exprs.add(re.compile(r'<.*select.*idref="'))
+
+    with open(datastream) as ds:
+        with tempfile.NamedTemporaryFile(mode='w', suffix='.xml') as new_ds:
+            for line in ds:
+                if any(x.search(line) for x in exprs):
+                    line = line.replace('selected="false"', 'selected="true"')
+                new_ds.write(line)
+            new_ds.flush()
+            util.log(f"Saving modified {datastream} with all rules being selected as"
+                     "{new_ds.name}")
+            yield Path(new_ds.name)
+
+
+def list_append_ansible_module_from_task(modules_list, task_keywords, allowed_modules):
+    updated_keywords = set()
+    for keyword in task_keywords:
+        keyword = keyword.replace("ansible.builtin.", "")
+        updated_keywords.add(keyword)
+    task_module = list(updated_keywords.intersection(allowed_modules))
+    if len(task_module) == 1:
+        module = str(task_module[0])
+        if modules_list.count(module) == 0:
+            # Module module is not yet in the list, append it.
+            modules_list.append(module)
+    else:
+        module = ''.join([x for x in list(updated_keywords) if x not in
+                          ansible_reserved_keywords])
+        results.report('fail', module, 'module is not allowed')
+        return 1
+    return 0
+
+
+def traverse_blocks(modules_list, task, allowed_modules):
+    block_set = set(['block'])
+    task_keywords = set(task.keys())
+
+    if len(task_keywords.intersection(block_set)) > 0:
+        for block_task in task['block']:
+            return traverse_blocks(modules_list, block_task, allowed_modules)
+    else:
+        if list_append_ansible_module_from_task(modules_list,
+                                                task_keywords,
+                                                allowed_modules):
+            util.log(f"This occurred in the following task:\n{task}\n")
+            return
+
+
+def check_allowed_modules(playbook_path, allowed_modules):
+    """
+    Collects all modules from Ansible playbook playbook_path and reports pass
+    if a module is found in the allowed_modules set, otherwise reports fail
+    and prints an offending Ansible module and task. Function also handles
+    Ansible block sections which can bundle multiple tasks.
+    Note: ansible playbook must be a valid playbook which means that
+    'ansible-playbook --syntax-check' must not report any errors for it.
+    """
+    with open(playbook_path, 'r') as yamlfile:
+        playbook = yaml.safe_load(yamlfile)
+    modules_list = []
+
+    util.log(f"Check Ansible playbook {playbook_path} contains only allowed modules")
+    for section in playbook:
+        if section['tasks'] is None:
+            # Playbook has no tasks.
+            return
```

Comment:
This aborts processing of other sections in the playbook. Maybe `continue` would be better?

If there is always only one section in every playbook (?), then don't use a `for` loop, a simple `section = next(playbook)` or `section = next(iter(playbook))` would more explicitly state there is only one section.